### PR TITLE
#156 Refactoring the error handler for canvas  request/responses

### LINKED ISSUE #156: Sending a better error response Object
The need for this issue came up as part of this PR bulk creating section see [here](https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/154#pullrequestreview-706497557), where response would contain multiple responses. The issue aim for generalizing the Error Response Object for all the Canvas API's

The Proposal is `{"statusCode":404,"errors":[{"message":"No response body was found.","failedInput":"403334"}]}`

The `failedInput` represents input that user provided, like for fetching the course name, it would be the course ID. Updating the course name it would be course new name. In case of creating new sections, it would be section name. The idea is that based on the request the respective feature API ( like getCourseName, putCourseName etc) will know how to deal with the errors

Note: This is code refactoring and get and put course name is not doing anything with returned response and I am going to keep it that way

### PR DESCRIPTION
**Pull request by pushyamig**
Fixes #156

### CODE REVIEW

File: ccm_web/client/src/utils/handleErrors.ts
Comment (ssciolla) [inline ccm_web/client/src/utils/handleErrors.ts:53]: So, if the client needs to use this as data, you're thinking they would just `JSON.parse`? I think one alternative would be to create a new `Error` subclass, maybe `DefaultError` or `APIError` that assigns the error data to an instance variable (through the constructor).

Class could look something like this:
```
class DefaultError extends Error {
  public name = 'DefaultError'
  errors: SingleError[] // types as you have them now

  constructor (errors: SingleError[]) {
    super('Received one or more errors associated with this request from Canvas')
    this.errors = errors
  }
}
```

and then you'd do this:
```
    default:
      errors = (JSON.parse(await resp.text()) as ErrorObj).errors
      throw new DefaultError(errors)
```

And then Chris or FE developer could do this in React components:
```
if ( setCourseNameError instanceof DefaultError ) {
  setCourseNameError.errors ...
}
```

File: ccm_web/server/src/api/api.interfaces.ts
Comment (ssciolla) [inline ccm_web/server/src/api/api.interfaces.ts:20]:
```suggestion

export interface APIErrorData {
```

File: ccm_web/server/src/api/api.interfaces.ts
Comment (ssciolla) [inline ccm_web/server/src/api/api.interfaces.ts:11]:
```suggestion

export interface APIErrorHandler {
```

Comment (ssciolla) [reply ccm_web/server/src/api/api.interfaces.ts:11]: I suggested combining this type with m`APIErrorPayload` elsewhere.

`Handler` sounds odd to me as a data type, I would expect it to be class.

**Review by ssciolla — COMMENTED** [review]
Releasing some comments/ideas

File: ccm_web/client/src/models/models.ts
Comment (lsloan) [inline ccm_web/client/src/models/models.ts:36]: WRT the problem described in #156…

_Posted by @pushyamig in https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/156#issue-945302924_
> The `failedInput` represents input that user provided, like for fetching the course name, it would be the course name.  Updating the course name it would be course new name. In case of creating new sections, it would be section name. The idea is that based on the request the respective feature API ( like getCourseName, putCourseName etc) will know how to deal with the errors.

Several of the API functions currently take one piece of data in its input payload.  Having this `failedInput` property contain a string that represents the input would probably be sufficient for most of those cases.  (That is, if you ignore instances where a number was the input, which is then converted to a string when it appears in `failedInput`.)  However, as we implement more complex functions, like creating many sections in one request or enrolling many users in one or more sections, it doesn't make sense to convert all that input to a string and put it in this single property.

Instead of a `failedInput` _string_ property, this needs to use `method` _string_, `request` _string_, and `payload` _object_ (or _"any"_-type, if that concept is supported) properties.

It could be used in the following cases…

* get course name error (getting course sections would be similar)
  `"method": "GET", "request": "/api/course/374350/name"`
* set course name error
  `"method": "PUT", "request": "/api/course/374350/name", "payload": {"newName": "test course for lsloan"}`
* create new sections error (I don't know the format of this request well, so this is approximate)
  `"method": "POST", "request": "/api/course/374350/sections", "payload": {"sections": ["sectionA", "sectionB"]}`

That way we'll know all of the parts of the request made that resulted in the error.

**Review by lsloan — CHANGES_REQUESTED** [review]

File: ccm_web/client/src/models/models.ts
Comment (pushyamig) [reply ccm_web/client/src/models/models.ts:35]: @ssciolla is this showing a linting error? since my IDE don't show this and you seems encounter it and asking for a change

Comment (ssciolla) [reply ccm_web/client/src/models/models.ts:35]: It's not, maybe it's just a carry-over from Python linting rules, I like to include a blank line between definitions. Not a huge deal, you can consider it optional.

File: ccm_web/server/src/api/api.utils.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.utils.ts]: Sorry for not removing the debug code. Should be gone

File: ccm_web/client/src/models/models.ts
Comment (pushyamig) [reply ccm_web/client/src/models/models.ts:43]: OK, I will make that change.

File: ccm_web/client/src/models/models.ts
Comment (pushyamig) [reply ccm_web/client/src/models/models.ts:37]: @lsloan For Multiple create sections look like as below. For concurrent sections creation we grab the info that is needs to send to the user, Like what section failed and what is the reason for failure. I am not sure what additional value by adding `method` and `url`. I am not saying No, I am trying to understand.
```
"errors": [
    {
      "failedInput": "sec1",
      "message": "Error(s) from Canvas:  Unprocessable Entity"
    },
    {
      "failedInput": "sec2",
      "message": "Error(s) from Canvas:  Unprocessable Entity"
    },
    
  ]
```

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts:38]: One weird thing i noticed with GetCourseName call is that if we assign `failedInput:undefined` the response object removes it. Why is this?

Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:38]: Hm, interesting. It might be because technically any string key not given a non-`undefined` value is `undefined`. I'd have to do some more research on this. It's not concerning to me though.

Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts:38]: But that kind of make the failedInput optional in the response object, should we make the response object more predictable or send null?

File: ccm_web/client/src/utils/handleErrors.ts
Comment (pushyamig) [reply ccm_web/client/src/utils/handleErrors.ts:53]: I am going by this approach

**Review by ssciolla — COMMENTED** [review]
@pushyamig, seems like you're working on things still. I'll let you finalize with @lsloan or @chrisrrowland, whatever you come up with will be okay with me, I will be flexible. No need to wait until I get back next week to merge.

File: ccm_web/server/src/api/api.service.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:38]: I'm okay either way; if you prefer `null`, we can do that, no worries.

File: ccm_web/client/src/models/models.ts
Comment (pushyamig) [reply ccm_web/client/src/models/models.ts]: Done

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts]: I made it null. and implemented the API structure as suggested

File: ccm_web/client/src/utils/handleErrors.ts
Comment (pushyamig) [reply ccm_web/client/src/utils/handleErrors.ts]: Done

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts]: I am sending null now instead of undefined since response object is taking it out for some reasons

**Review by ssciolla — APPROVED** [review]
This is a tricky problem. While this may not be perfect, I do think it gets us closer to effectively capturing errors in an extensible manner. I do think we will want to do more on `handleErrors` and the associated `Error` subclasses, and probably handle or customize the response data the DTO validation generates. All this can be done later.

I did a little testing, I didn't see any major red flags. We can continue to iterate on this. I'm approving.

I've left a few comments, feel free to handle as you see fit, no need to re-request a review.

File: ccm_web/server/src/api/api.interfaces.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.interfaces.ts:11]: I am keeping it the same

File: ccm_web/server/src/api/api.interfaces.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.interfaces.ts:14]: Yeah, I think `failedInput` is going to be evolve as we start adding API endpoint, I feel it good the you are thinking that far. But we couldn't possibly think all the future option to make it perfect. We can incrementally make it better.

File: ccm_web/server/src/api/api.utils.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.utils.ts]: Done

File: ccm_web/client/src/models/models.ts
Comment (lsloan) [reply ccm_web/client/src/models/models.ts]: I'm suggesting that it would be helpful to have all of the information about the request included in the error message.  That will be helpful for debugging problems with the API.

Comment (pushyamig) [reply ccm_web/client/src/models/models.ts]: @lsloan based on the comments [here](https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/157/files#r670803544) the error response object is going to be an evolving thing. I think we can discuss further what would be useful. 

Currently the rational used what is useful info the FE needs to the user and based on that the object is framed.

Comment (lsloan) [reply ccm_web/client/src/models/models.ts]: @pushyamig, I don't have anything else to add to the discussion other than what I wrote above.

The information I recommended to be included in the error data are what I would consider a minimum requirement.  This error data could also be logged for debugging and support purposes and would include almost everything that needs to be known to reproduce the request.

You mentioned something about over engineering recently.  I don't know whether it was about this suggestion.  I agree that can be a problem, but under engineering is also a problem.

**Pull request merged**
